File: tests/Kotlin/MyApplication1/app/src/main/java/com/example/myapplication1/MainViewModel.kt
// SPDX-FileCopyrightText: 2024 Rifat Hasan <atunutemp1@gmail.com>
//
// SPDX-License-Identifier: MIT

package com.example.myapplication1

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

class MainViewModel : ViewModel() {
    val nomuna = Nomuna()
    val inputHolder = InputHolder()

    init {
        nomuna.nimMain()
        nomuna.noop()
        nomuna.extraNoOp()
        nomuna.printCond(inputHolder.addInt1 == "")
        nomuna.printCond(inputHolder.addInt1 != "")
        nomuna.takeChar('a')
        nomuna.printStr("nim")
        nomuna.printStr("hello ñíℳ")
        nomuna.print2Str("Hello", "World!")
        var direction = Nomuna.Direction.SOUTH
        nomuna.printDirectionRawValue(direction)
        direction = nomuna.getDirection("south")
        Log.d("nomuna", "Direction: $direction, value: ${direction.ordinal}")
        val gameState = Nomuna.GameState.GAME_OVER
        Log.d("nomuna", "Game State: $gameState, value: ${gameState.intVal}")
        val newGameState = nomuna.togglePause(gameState)
        Log.d("nomuna", "Game State: $newGameState, value: ${newGameState.intVal}")
        var statusCode = nomuna.authenticate("user1")
        Log.d("nomuna", "Status code: $statusCode, value: ${statusCode.intVal}")
        statusCode = nomuna.setGameState("user", Nomuna.GameState.GAME_OVER)
        Log.d("nomuna", "set Game State result: $statusCode, value: ${statusCode.intVal}")
        val newUser = nomuna.requestPermission(Nomuna.FilePermission.WRITE)
        Log.d("nomuna", "$newUser has permission: ${Nomuna.FilePermission.WRITE}")
    }

    val retCardUiState =
        ReturnCardUiState(
            nomuna.constRet(),
            nomuna.constRetBool(),
            nomuna.constRetFloat(),
            nomuna.constRetChar(),
            nomuna.constRetStr(),
            nomuna.constRetUnicodeStr(),
        )

    private var _addCardUiState = MutableStateFlow(AddCardUiState())
    val addCardUiState: StateFlow<AddCardUiState> = _addCardUiState.asStateFlow()

    private var _inputCardUiState =
        MutableStateFlow(
            InputCardUiState(
                oppositeDirection = nomuna.getOpposite(Nomuna.Direction.entries.first()).name,
            ),
        )
    val inputCardUiState: StateFlow<InputCardUiState> = _inputCardUiState.asStateFlow()

    inner class InputHolder : InputContainer {
        override var addInt1 by mutableStateOf("")
            private set
        override var addInt2 by mutableStateOf("")
            private set
        override var addDouble1 by mutableStateOf("")
            private set
        override var addDouble2 by mutableStateOf("")
            private set
        override var addFloat1 by mutableStateOf("")
            private set
        override var addFloat2 by mutableStateOf("")
            private set
        override var sayHelloInput by mutableStateOf("")
            private set

        override fun updateIntNum1(num: String) {
            if (!verifyIntOrEmpty(num)) {
                return
            }
            addInt1 = num
            performAddInt()
        }

        override fun updateIntNum2(num: String) {
            if (!verifyIntOrEmpty(num)) {
                return
            }
            addInt2 = num
            performAddInt()
        }

        override fun updateDoubleNum1(num: String) {
            if (!verifyDoubleOrEmpty(num)) {
                return
            }
            addDouble1 = num
            performAddDouble()
        }

        override fun updateDoubleNum2(num: String) {
            if (!verifyDoubleOrEmpty(num)) {
                return
            }
            addDouble2 = num
            performAddDouble()
        }

        override fun updateFloatNum1(num: String) {
            if (!verifyFloatOrEmpty(num)) {
                return
            }
            addFloat1 = num
            performAddFloat()
        }

        override fun updateFloatNum2(num: String) {
            if (!verifyFloatOrEmpty(num)) {
                return
            }
            addFloat2 = num
            performAddFloat()
        }

        override fun updateStrInput(str: String) {
            sayHelloInput = str
            performSayHello()
        }

        override fun updateSelectedDirection(direction: String) {
            performDirectionSelection(direction)
        }
    }

    private fun verifyIntOrEmpty(num: String): Boolean = num.toIntOrNull() != null || num.isEmpty()

    private fun verifyDoubleOrEmpty(num: String): Boolean = num.toDoubleOrNull() != null || num.isEmpty()

    private fun verifyFloatOrEmpty(num: String): Boolean =
        (
            num.toFloatOrNull() != null &&
                (if (num.contains('.')) num.substring(num.indexOf('.') + 1).length <= 2 else true)
        ) ||
            num.isEmpty()

    private fun performAddInt() {
        val num1 = inputHolder.addInt1.toIntOrNull() ?: 0
        val num2 = inputHolder.addInt2.toIntOrNull() ?: 0
        _addCardUiState.update { currentState ->
            currentState.copy(addIntRes = nomuna.addIntNum(num1, num2).toString())
        }
    }

    private fun performAddDouble() {
        val num1 = inputHolder.addDouble1.toDoubleOrNull() ?: 0.0
        val num2 = inputHolder.addDouble2.toDoubleOrNull() ?: 0.0
        _addCardUiState.update { currentState ->
            currentState.copy(addDoubleRes = "%.4f".format(nomuna.addDouble(num1, num2)))
        }
    }

    private fun performAddFloat() {
        val num1 = inputHolder.addFloat1.toFloatOrNull() ?: 0.0f
        val num2 = inputHolder.addFloat2.toFloatOrNull() ?: 0.0f
        _addCardUiState.update { currentState ->
            currentState.copy(addFloatRes = "%.2f".format(nomuna.addFloat(num1, num2)))
        }
    }

    private fun performSayHello() {
        _inputCardUiState.update { currentState ->
            currentState.copy(sayHelloOutput = nomuna.sayHello(inputHolder.sayHelloInput))
        }
    }

    private fun performDirectionSelection(direction: String) {
        val selectedDirection = Nomuna.Direction.entries.first { it.name == direction }
        val oppositeDirection = nomuna.getOpposite(selectedDirection)
        _inputCardUiState.update { currentState ->
            currentState.copy(oppositeDirection = oppositeDirection.name)
        }
    }
}

interface InputContainer {
    val addInt1: String

    fun updateIntNum1(num: String)

    val addInt2: String

    fun updateIntNum2(num: String)

    val addDouble1: String

    fun updateDoubleNum1(num: String)

    val addDouble2: String

    fun updateDoubleNum2(num: String)

    val addFloat1: String

    fun updateFloatNum1(num: String)

    val addFloat2: String

    fun updateFloatNum2(num: String)

    val sayHelloInput: String

    fun updateStrInput(str: String)

    fun updateSelectedDirection(direction: String)
}

File: tests/C#/App1/App1/ViewModels/MainViewModel.cs
// SPDX-FileCopyrightText: 2024 Rifat Hasan <atunutemp1@gmail.com>
//
// SPDX-License-Identifier: MIT

using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using NomunaLib;
using Windows.Globalization.NumberFormatting;

namespace App1.ViewModels;

public partial class MainViewModel : ObservableRecipient
{
    [ObservableProperty]
    private int intRetVal;

    [ObservableProperty]
    private bool boolRetVal;

    [ObservableProperty]
    private double doubleRetVal;

    [ObservableProperty]
    private char charRetVal;

    [ObservableProperty]
    private string stringRetVal;

    [ObservableProperty]
    private string unicodeStringRetVal;

    [ObservableProperty]
    private string addInt1;

    [ObservableProperty]
    private string addInt2;

    [ObservableProperty]
    private string addIntRes;

    [ObservableProperty]
    private string addDouble1;

    [ObservableProperty]
    private string addDouble2;

    [ObservableProperty]
    private string addDoubleRes;

    [ObservableProperty]
    private string addFloat1;

    [ObservableProperty]
    private string addFloat2;

    [ObservableProperty]
    private string addFloatRes;

    [ObservableProperty]
    private string sayHelloInput;

    [ObservableProperty]
    private string sayHelloOutput;

    [ObservableProperty]
    private ObservableCollection<Nomuna.Direction> directions = new(
        Enum.GetValues<Nomuna.Direction>()
    );

    [ObservableProperty]
    private Nomuna.Direction selectedDirection;

    public string OppositeDirection
    {
        get
        {
            var oppositeDir = Nomuna.GetOpposite(selectedDirection);
            return oppositeDir.ToString();
        }
    }

    private readonly Lazy<DecimalFormatter> _intFormatter =
        new(() =>
        {
            var formatter = new DecimalFormatter { FractionDigits = 0 };
            var rounder = new IncrementNumberRounder
            {
                RoundingAlgorithm = RoundingAlgorithm.RoundDown
            };
            formatter.NumberRounder = rounder;
            return formatter;
        });
    public DecimalFormatter IntFormatter => _intFormatter.Value;

    private readonly Lazy<DecimalFormatter> _doubleFormatter =
        new(() =>
        {
            var formatter = new DecimalFormatter { FractionDigits = 4 };
            var rounder = new IncrementNumberRounder { Increment = 0.0001 };
            formatter.NumberRounder = rounder;
            return formatter;
        });
    public DecimalFormatter DoubleFormatter => _doubleFormatter.Value;

    private readonly Lazy<DecimalFormatter> _floatFormatter =
        new(() =>
        {
            var formatter = new DecimalFormatter { FractionDigits = 2 };
            var rounder = new IncrementNumberRounder { Increment = 0.01 };
            formatter.NumberRounder = rounder;
            return formatter;
        });
    public DecimalFormatter FloatFormatter => _floatFormatter.Value;

    public MainViewModel()
    {
        intRetVal = Nomuna.ConstRet();
        boolRetVal = Nomuna.ConstRetBool();
        doubleRetVal = Nomuna.ConstRetFloat();
        charRetVal = Nomuna.ConstRetChar();
        stringRetVal = Nomuna.ConstRetStr();
        unicodeStringRetVal = Nomuna.ConstRetUnicodeStr();
        addInt1 =
            addInt2 =
            addDouble1 =
            addDouble2 =
            addFloat1 =
            addFloat2 =
            sayHelloInput =
            sayHelloOutput =
                "";
        addIntRes = addDoubleRes = addFloatRes = "0";
        selectedDirection = Nomuna.Direction.North;
        Nomuna.PrintCond(addIntRes == "0");
        Nomuna.PrintCond(addIntRes != "0");
        Nomuna.TakeChar('a');
        Nomuna.PrintStr("nim");
        Nomuna.PrintStr("hello ñíℳ");
        Nomuna.Print2Str("Hello", "World!");
        Nomuna.Direction direction = Nomuna.Direction.South;
        Nomuna.PrintDirectionRawValue(direction);
        direction = Nomuna.GetDirection("south");
        Console.WriteLine($"Direction: {direction}, value: {(int)direction}");
        Nomuna.GameState gameState = Nomuna.GameState.Game_over;
        Console.WriteLine($"Game State: {gameState}, value: {((int)gameState)}");
        Nomuna.GameState newGameState = Nomuna.TogglePause(gameState);
        Console.WriteLine($"Game State: {newGameState}, value: {((int)newGameState)}");
        Nomuna.HttpStatusCode statusCode = Nomuna.Authenticate("user1");
        Console.WriteLine($"Status code: {statusCode}, value: {(int)statusCode}");
        statusCode = Nomuna.SetGameState("user", Nomuna.GameState.Game_over);
        Console.WriteLine($"set Game State result: {statusCode}, value: {(int)statusCode}");
        string newUser = Nomuna.RequestPermission(Nomuna.FilePermission.Write);
        Console.WriteLine(
            $"{newUser} has permission: {Nomuna.FilePermission.Write}, value: 0x{(int)Nomuna.FilePermission.Write:x2}"
        );
        Nomuna.FilePermission permission = Nomuna.GetLeastPriviledgedPermission();
        Console.WriteLine($"Least priviledged permission: {permission}, value: {(int)permission}");
    }

    [RelayCommand]
    private void PerformAddInt()
    {
        int i;
        var num1 = int.TryParse(AddInt1, out i) ? i : 0;
        var num2 = int.TryParse(AddInt2, out i) ? i : 0;
        var sum = Nomuna.AddIntNum(num1, num2);
        AddIntRes = sum.ToString();
    }

    [RelayCommand]
    private void PerformAddDouble()
    {
        double d;
        var num1 = double.TryParse(AddDouble1, out d) ? d : 0;
        var num2 = double.TryParse(AddDouble2, out d) ? d : 0;
        var sum = Nomuna.AddDouble(num1, num2);
        sum = Math.Round(sum, 4);
        AddDoubleRes = sum.ToString();
    }

    [RelayCommand]
    private void PerformAddFloat()
    {
        float f;
        var num1 = float.TryParse(AddFloat1, out f) ? f : 0;
        var num2 = float.TryParse(AddFloat2, out f) ? f : 0;
        var sum = Nomuna.AddFloat(num1, num2);
        AddFloatRes = sum.ToString();
    }

    [RelayCommand]
    private void PerformSayHello()
    {
        SayHelloOutput = Nomuna.SayHello(SayHelloInput);
    }

    partial void OnSelectedDirectionChanged(Nomuna.Direction value)
    {
        OnPropertyChanged(nameof(OppositeDirection));
    }
}
